tests/Unit/Pypp/PyppFundamentals.hpp

```diff
@@ -0,0 +1,322 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines classes for converting to and from Python objects
+
+#pragma once
+
+#include <Python.h>
+#include <array>
+#include <cstddef>
+#include <initializer_list>
+#include <stdexcept>
+#include <type_traits>
+#include <vector>
+
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Requires.hpp"
+#include "Utilities/TypeTraits.hpp"
+
+namespace pypp {
+/// \cond
+using None = void*;
+
+template <typename T, typename = std::nullptr_t>
+struct ToPyObject;
+
+template <typename T>
+PyObject* to_py_object(const T& t) {
+  PyObject* value = ToPyObject<T>::convert(t);
+  if (value == nullptr) {
+    throw std::runtime_error{"Failed to convert argument."};
+  }
+  return value;
+}
+
+template <typename T, typename = std::nullptr_t>
+struct FromPyObject;
+
+template <typename T>
+T from_py_object(PyObject* t) {
+  return FromPyObject<T>::convert(t);
+}
+/// \endcond
+
+/// Create a python tuple from Args
+///
+/// \tparam Args the types of the arguments to be put in the tuple (deducible)
+/// \param t the arguments to put into the tuple
+/// \return PyObject* containing a Python tuple
+template <typename... Args>
+PyObject* make_py_tuple(const Args&... t) {
+  PyObject* py_tuple = PyTuple_New(sizeof...(Args));
+  int entry = 0;
+  (void)std::initializer_list<char>{[&entry, &py_tuple](const Args& arg) {
```

Comment:
With the compilation errors, you might have to put extra parens around the lambda. Could you try that and also squash and rebase on develop?

Reply:
GCC is known to have some bugs with expanding parameter pack expressions containing lambdas.  [This one](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=47226) looks fairly similar to the code here, so if you still get errors after trying some variations that's likely the problem.  (There is a fix in version 8 but it looks unlikely to be backported.)

On an unrelated topic, the example in the bug contains `void slurp(...) { }` if you're looking for "swallow" alternatives.

Reply:
Assigning the lambda beforehand and using a generic lambda seems to work, but with a compiler warning that `add_entry` is unused. (Using a generic lambda and inlining it doesn't work). 
```  int entry = 0;
  auto add_entry = [&entry, &py_tuple](const auto& arg) {
    PyObject* value = to_py_object(arg);
    PyTuple_SetItem(py_tuple, entry, value);
    entry++;
    return '0';
  };
  (void)std::initializer_list<char>{add_entry(t)...};
```

Reply:
Ah, this is because sometimes the pack is empty, just add a `(void)add_entry;` line, we've had to do this elsewhere. Also, could you const the lambda?